dvc/external_repo.py

```diff
@@ -123,21 +123,40 @@ def _set_cache_dir(self):
 
         self.cache.local.cache_dir = cache_dir
 
-    def _set_upstream(self):
-        # check if the URL is local and no default remote is present
-        # add default remote pointing to the original repo's cache location
+    def _fix_upstream(self):
         if os.path.isdir(self.url):
-            if not self.config["core"].get("remote"):
-                src_repo = Repo(self.url)
-                try:
-                    cache_dir = src_repo.cache.local.cache_dir
-                finally:
-                    src_repo.close()
-
-                self.config["remote"]["auto-generated-upstream"] = {
-                    "url": cache_dir
-                }
-                self.config["core"]["remote"] = "auto-generated-upstream"
+            upstream_name = self.config["core"].get("remote")
+            if upstream_name:
+                self._fix_local_remote(upstream_name)
+            else:
+                self._add_upstream()
+
+    def _fix_local_remote(self, remote_name):
+        # Keep the relative path for a local remote relative
+        # to the original repo.
+        src_repo = Repo(self.url)
+        try:
+            new_remote = self.config["remote"][remote_name]
+            old_remote = src_repo.config["remote"][remote_name]
+            # NOTE: it is assumed that only local+relative urls
+            # will be different from their old version. Revise this
+            # test if that stops being true.
+            if new_remote["url"] != old_remote["url"]:
+                new_remote["url"] = old_remote["url"]
+        finally:
+            src_repo.close()
+
+    def _add_upstream(self):
+        # Fill the empty upstream entry with a new remote pointing to the
+        # original repo's cache location.
+        src_repo = Repo(self.url)
+        try:
+            cache_dir = src_repo.cache.local.cache_dir
+        finally:
+            src_repo.close()
```

Comment:
Let's move
```
src_repo = Repo(self.url)
try:
     ...
finally:
    src_repo.close()
```
to _fix_upstream, it will deduplicate some code.

Reply:
I thought about that, but I was unsure because of how `_set_upstream` still sets an upstream even if obtaining the cache dir fails. But since in that case it will set the url to `None` I guess it doesn't matter?